Add wrappers for scrypt key derivation and password hashing

hsivonen opened the pull request:
Closes [issue 21](https://github.com/stef/pysodium/issues/21).

stef commented:
awesome! please add testcases for all new functions to test/test_pysodium.py and if they pass i'll merge this.

thanks a lot!

bgaifullin commented on test/test_pysodium.py:
```diff
@@ -121,5 +121,55 @@ def test_crypto_blake2b(self):
         self.assertEqual(pysodium.crypto_generichash_blake2b_salt_personal(message, personal = key, key = key), binascii.unhexlify(b'5a0b3db4bf2dab71485211447fc2014391228cc6c1acd2f3031050a9a32ca407'))
 
+    def test_crypto_pwhash_scryptsalsa208sha256(self):
+        passwd =       b'Correct Horse Battery Staple'
```
Multiple spaces after '=', please fix.

bgaifullin commented on test/test_pysodium.py:
```diff
@@ -121,5 +121,55 @@ def test_crypto_blake2b(self):
+        self.assertNotEqual(key, pysodium.crypto_pwhash_scryptsalsa208sha256(32, other_passwd, salt, ops_limit, mem_limit))
+
+    def test_crypto_pwhash_scryptsalsa208sha256_str(self):
+        passwd =       b'Correct Horse Battery Staple'
```
multiple spaces after '='

bgaifullin commented on test/test_pysodium.py:
```diff
@@ -121,5 +121,55 @@ def test_crypto_blake2b(self):
+        self.assertNotEqual(storage_string, pysodium.crypto_pwhash_scryptsalsa208sha256_str(passwd, ops_limit, mem_limit), "Each call should compute a new random salt.")
+
+    def test_crypto_pwhash_scryptsalsa208sha256_str_verify(self):
+        passwd =       b'Correct Horse Battery Staple'
```
multiple spaces after '='

hsivonen commented:
@bgaifullin:

I've removed the multiple spaces even if it makes it less obvious visually what the different between `passwd` and `other_passwd` and `salt` and `other_salt` is.

@stef:
> awesome! please add testcases for all new functions to test/test_pysodium.py and if they pass i'll merge this.

It looks like instead of merging this PR, you merged [pull request 20](https://github.com/stef/pysodium/pull/20) instead even though the code there is incorrect on multiple counts. :-(

Should I expect PR 20 to be backed out and this PR to be merged in the current form of this PR or should I expect to have to reformulate this PR as a fix on top of that one?

hsivonen commented:
> Should I expect PR 20 to be backed out and this PR to be merged in the current form of this PR or should I expect to have to reformulate this PR as a fix on top of that one?

I went ahead and did the latter.

hsivonen commented:
Note that I intentionally did not provide default values for `memlimit` and `opslimit` to force the caller to think about them as the caller should.

bgaifullin commented on test/test_pysodium.py:
```diff
@@ -121,18 +121,55 @@ def test_crypto_blake2b(self):
+        passwd = b'Correct Horse Battery Staple'
+
+        # Use very small limits to avoid burning resources in CI
+        mem_limit = 32*1024
```
space between *

bgaifullin commented on test/test_pysodium.py:
```diff
@@ -121,18 +121,55 @@ def test_crypto_blake2b(self):
+
+        storage_string = pysodium.crypto_pwhash_scryptsalsa208sha256_str(passwd, ops_limit, mem_limit)
+        self.assertTrue(storage_string.startswith(pysodium.crypto_pwhash_scryptsalsa208sha256_STRPREFIX))
+        self.assertFalse(b'\x00' in storage_string)
```
self.assertIn

bgaifullin commented on test/test_pysodium.py:
```diff
@@ -121,18 +121,55 @@ def test_crypto_blake2b(self):
+        other_passwd = b'correct horse battery staple'
+        
+        # Use very small limits to avoid burning resources in CI
+        mem_limit = 32*1024
```
space around *

bgaifullin commented on test/test_pysodium.py:
```diff
@@ -121,18 +121,55 @@ def test_crypto_blake2b(self):
+        other_salt = binascii.unhexlify(b'4206baae5578933d7cfb315b1c257cc7af162965a91a74ccbb1cfa1d747eb692')
+        
+        # Use very small limits to avoid burning resources in CI
+        mem_limit = 32*1024
```
space around

hsivonen commented:
@stef, could you, please, clarify what the expectations are here?

I have added tests and I have addressed @bgaifullin's review comments. The code itself has been correct all along. Yet, instead of this PR getting merged, first incorrect code gets merged from [PR 20](https://github.com/stef/pysodium/pull/20) and now instead of merging this PR, which would fix multiple bugs introduced in PR 20, [PR 25](https://github.com/stef/pysodium/pull/25) got priority even though it only fixed one problem with PR 20 and even the fix is bad (`.raw[:-1]` instead of `.value').

The code introduced in PR 20 continues to be dangerously incorrect in its salt handling, fails to take the output length in the first function from the caller and lacks the `crypto_pwhash_scryptsalsa208sha256_STRPREFIX` constant. The proposed changes here have been correct since before PR 20 was merged.

stef commented:
thanks for the fixes

hsivonen commented:
Thank you for the merge.
